src/codec.rs

```diff
@@ -0,0 +1,474 @@
+//! of VDAF messages in TLS syntax, as specified in
+//! [RFC 8446, Section 3](https://datatracker.ietf.org/doc/html/rfc8446#section-3). We provide
+//! traits that can be implemented on values that need to be encoded or decoded, as well as utility
+//! functions for encoding sequences of values.
+//!
+//! TLS syntax supports
+//! [variable length vectors](https://datatracker.ietf.org/doc/html/rfc8446#section-3.4) of objects.
+//! The wire encoding is to write the number of objects in the vector and then the concatenated
+//! objects. TLS also defines an [`opaque`](https://datatracker.ietf.org/doc/html/rfc8446#section-3.2)
+//! type for opaque bytes. The `encode_items_*` and `decode_items_*` functions encode or decode a
+//! sequence of values of types which implement [`Encode`] or [`Decode`], where the the length
+//! marker is the number of *items*. The `encode_items_opaque_*` and `decode_items_opaque_*`
+//! functions encode or decode a sequence of values which implement [`Encode`] or [`Decode`] as
+//! opaque vectors, so the length marker is the number of *bytes* occupied by the encoded
+//! representation of the concatenated values. Finally, the `encode_opaque_*` and `decode_opaque_*`
+//! functions allow encoding and decoding sequences of [`u8`] as opaque vectors.
+//!
+//! It is very tempting to try to express the {encode,decode}_items_opaque_*,
+//! {encode,decode}_opaque_*, {encode,decode}_items_* and corresponding decode_* functions
+//! generically over the type of the length prefix. Unfortunately, const generics and const
+//! expressions are not mature enough to do this gracefully, and since we only have to provide a
+//! small number of specializations for u8, u16 and maybe u24, we do it by hand.
+
+use std::{
+    convert::TryInto,
+    io::{Cursor, Read},
+    mem::size_of,
+};
+
+/// Describes how to decode an object from a byte sequence.
+pub trait Decode<D>: Sized {
+    /// Type of errors returned by the [`decode`] method.
+    type Error;
+
+    /// Read and decode an encoded object from `bytes`. `decoding_parameter` provides details of the
+    /// wire encoding such as lengths of different portions of the message.
+    fn decode(decoding_parameter: &D, bytes: &mut Cursor<&[u8]>) -> Result<Self, Self::Error>;
+
+    /// Convenience method to get decoded value.
+    fn get_decoded(decoding_parameter: &D, bytes: &[u8]) -> Result<Self, Self::Error> {
+        let mut cursor = Cursor::new(bytes);
+        Self::decode(decoding_parameter, &mut cursor)
+    }
+}
+
+/// Describes how to encode objects into a byte sequence.
+pub trait Encode {
+    /// Append the encoded form of this object to the end of `bytes`, growing the vector as needed.
+    fn encode(&self, bytes: &mut Vec<u8>);
+
+    /// Convenience method to get encoded value.
+    fn get_encoded(&self) -> Vec<u8> {
+        let mut ret = Vec::new();
+        self.encode(&mut ret);
+        ret
+    }
+}
+
+impl Decode<()> for u8 {
```

Comment:
For the `Encode`/`Decode` implementations on primitive numeric types, consider use of the [byteorder](https://docs.rs/byteorder/1.4.3/byteorder/) crate -- I think all of the implementations would be reduced to a one-liner.